client/allocrunner/taskrunner/envoybootstrap_hook.go

```diff
@@ -231,12 +257,32 @@ func (h *envoyBootstrapHook) Prestart(ctx context.Context, req *interfaces.TaskP
 	return nil
 }
 
-func buildEnvoyAdminBind(alloc *structs.Allocation, taskName string) string {
+// buildEnvoyAdminBind determines a unique port for use by the envoy admin
+// listener.
+//
+// In bridge mode, if multiple sidecars are running, the bind addresses need
+// to be unique within the namespace, so we simply start at 19000 and increment
+// by the index of the task.
+//
+// In host mode, use the port provided through the service definition, which can
+// be a port chosen by Nomad.
+func buildEnvoyAdminBind(alloc *structs.Allocation, serviceName, taskName string) string {
+	tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
 	port := envoyBaseAdminPort
-	for idx, task := range alloc.Job.LookupTaskGroup(alloc.TaskGroup).Tasks {
-		if task.Name == taskName {
-			port += idx
-			break
+	switch tg.Networks[0].Mode {
+	case "host":
+		for _, service := range tg.Services {
+			if service.Name == serviceName {
+				_, port = tg.Networks.Port(service.PortLabel)
+				break
+			}
+		}
```

Comment:
Since this is bound on `localhost` we don't have to reserve a port and instead could bind to `127.0.0.1:0`, let the OS pick a free port, and have Envoy write that to a file with `-admin-address-path`: https://www.envoyproxy.io/docs/envoy/latest/operations/cli#cmdoption-admin-address-path